README.md
# python-libnmap

## Code status

![preflight-check](https://github.com/savon-noir/python-libnmap/workflows/Preflight%20Check/badge.svg)
[![Coverage Status](https://coveralls.io/repos/github/savon-noir/python-libnmap/badge.svg?branch=master)](https://coveralls.io/github/savon-noir/python-libnmap?branch=master)
[![License](https://img.shields.io/badge/License-Apache%202.0-blue.svg)](https://opensource.org/licenses/Apache-2.0)

## Use cases

libnmap is a python library enabling python developers to manipulate nmap process and data.

libnmap is what you were looking for if you need to implement the following:

- automate or schedule nmap scans on a regular basis
- manipulate nmap scans results to do reporting
- compare and diff nmap scans to generate graphs
- batch process scan reports
- ...

The above uses cases will be easy to implement with the help of the libnmap modules.

## libnmap modules

The lib currently offers the following modules:

- **process**: enables you to launch nmap scans
- **parse**: enables you to parse nmap reports or scan results (only XML so far) from a file, a string,...
- **report**: enables you to manipulate a parsed scan result and de/serialize scan results in a json format
- **diff**: enables you to see what changed between two scans
- **common**: contains basic nmap objects like NmapHost and NmapService. It is to note that each object can be "diff()ed" with another similar object.
- **plugins**: enables you to support datastores for your scan results directly in the "NmapReport" object. from report module:
    - mongodb: insert/get/getAll/delete
    - sqlalchemy: insert/get/getAll/delete
    - aws s3: insert/get/getAll/delete (not supported for python3 since boto is not supporting py3)
    - csv: todo (easy to implement)
    - elastic search: todo

## Documentation

All the documentation is available on [read the docs](https://libnmap.readthedocs.org). This documentation contains small code samples that you directly reuse.

## Dependencies

libnmap has by default no dependencies.

The only additional python modules you'll have to install depends if you wish to use libnmap to store reports on an exotic data store via libnmap's independents plugins.

Below the list of optional dependencies:

- [sqlalchemy](https://github.com/zzzeek/sqlalchemy) (+the driver ie:MySQL-python)
- [pymongo](https://github.com/mongodb/mongo-python-driver/)
- [boto](https://github.com/boto/boto)

## Python Support

The libnmap code is tested against the following python interpreters:

- Python 2.6
- Python 2.7
- Python 3.3
- Python 3.4
- Python 3.5
- Python 3.6
- Python 3.7
- Python 3.8

## Install

You can install libnmap via pip:

```bash
ronald@brouette:~$ pip install python-libnmap
```

or via git and dist utils (à l'ancienne):

```bash
ronald@brouette:~$ git clone https://github.com/savon-noir/python-libnmap.git
ronald@brouette:~$ cd python-libnmap
ronald@brouette:~$ python setup.py install
```

or via git and pip:

```bash
ronald@brouette:~$ git clone https://github.com/savon-noir/python-libnmap.git
ronald@brouette:~$ cd python-libnmap
ronald@brouette:~$ pip install .
```

## Examples

Some codes samples are available in the examples directory or in the [documentation](https://libnmap.readthedocs.org).

Among other example, you notice an sample code pushing nmap scan reports in an ElasticSearch instance and allowing you to create fancy dashboards in Kibana like the screenshot below:

![Kibanane](https://github.com/savon-noir/python-libnmap/blob/es/examples/kibanalibnmap.png)

## Contributors

Mike @bmx0r Boutillier for S3 and SQL-Alechemy plugins and for the constructive critics. Thanks!


CHANGES.txt
v0.7.1, 22/11/2020 -- code clean-up + fix for CVE-2019-1010017
v0.7.0, 28/02/2016 -- A few bugfixes
                    - fixe of endless loop in Nmap.Process. Fix provided by @rcarrillo, many thanks!
v0.6.3, 18/08/2015 -- Merged pull requests for automatic pypi upload, thanks @bmx0r
v0.6.2, 03/01/2015 -- Fixed a few bugs
                    - bulk fix for issues 37, 41, 42, 43, 44, 46
                    - added cpe_list method
                    - added elasticsearch example code
v0.6.1, 29/06/2014 -- Added full support for python 3.X
                    - python now supports python 2.6, 2.7, 3.3, 3.4
v0.5.1, 26/05/2014 -- Added basic API for class CPE
                    - interface similar to python-cpe for more
                      advanced usage of CPE, I strongly recommend you
                      to use python-cpe. Nice code, good doc.
v0.5.0, 17/05/2014 -- Rewrite of NmapProcess
                    - removed Threads to read stdout/stderr
                    - replaced by Process
                    - use of semaphore to not consume CPU while looping
                    - added NmapTask class
                    - added NmapProcess.current_task
                    - added NmapProcess.tasks list
                    - fixed bug in NmapProcess.state
                    - note NmapProcess.is_alive() is not recommended
v0.4.9, 14/05/2014 -- Fix for issue 28 and added code samples
v0.4.8, 05/05/2014 -- Changes in OS fingerprint data API
                    - NmapHost now holds its OS fingerprint data in NmapHost.os
                      (NmapOSFingerpring object)
                    - fingerprint is now a property which means you have to call
                      it without () NmapHost.os.fingerprint and should be called
                      directly from NmapHost.os
                    - NmapHost.os.fingerprints return an array of os
                      fingerprints (strings)
                    - NmapHost.os.fingerprint return a concatenated string of
                      all fingerprints
                    - Fingerprints data are now accessible via
                      NmapHost.os.osmatches which returns a list of NmapOSMatch
                      objects
                    - NmapOSMatch objects might contain a list of NmapOSClass
                      objects matching with it
                    - NmapOSClass objects might contain a list of CPE object
                      related to the os class (CPE class will be improved and
                      API enriched)
                    - TODO: finalize API doc and create all related unit tests
v0.4.7, 03/05/2014 -- minor fix for issue25
                    - fixed exception when optional service tag is not present
                      in <port> tag
                    - added support for <owner> if present in <port>:
                      accessible via NmapService.owner
v0.4.6, 06/04/2014 -- minor fix
                    - corrected missing incomplete parameter on parse_fromfile
                      and parse_fromstring
                    - added support to run scan in background with sudo support
                      added NmapProcess.sudo_run_background()
                    - fixed issue with run() blocking when an error triggered
                      during the scan
v0.4.5, 06/04/2014 -- minor fixes and botox injections
                    - Added "incomplete" argument in NmapReport.parse()
                      in order to enable parsing of incomplete or interrupted
                      nmap scans. Could be useful to use with a background
                      scan to parse incomplete data blocks from callback
                      function (thanks @Sibwara for the idea).
                    - Fixed bug when NmapReport.summary is empty
                    - added NmapReport.endtimestr
                    - remplaced ElementTree by cElementTree (performance)
v0.4.4, 04/04/2014 -- Bugs and features requests
                    - added support for tunnel attribute from <service> tag
                    - added support for servicefp (service fingerprint) in
                      attributes from <port><service> tag
                    - added support for reasons attributes from <port> tag
                    - added support for extraports/extrareasons tags
                    - corrected bug in serialization: missing extra data
                      (pull request from DougRoyal)
v0.4.3, 14/03/2014 -- Lots of bug corrections - Warning small changes in API:
                      - fix issue#14: better scripts parsing
                      - API change for NmapService.scripts_results:
                          NmapService.scripts_results is now a property
                          NmapService.scripts_results return an array of scripts results
                  
                      - fix issue#9 address field not correcly parsed:
                        MAC Address would erase an ipv4 address type.
                      - added new properties in hosts object API:
                          NmapHost.ipv4, NmapHost.ipv6, NmapHost.mac
                      - NmapHost.address property returns the IPv4 or IPv6 or MAC
                        in that preference order. Use specific calls for determinists
                        results
                      - fix API issue#10: os_ports_used
v0.4.2, 26/12/2013 -- Bug corrected: issue 8
                      There is no guarantee that "finished" or "runstats" will be
                      received by event parser of process.py.
                      Summary functions are now flagged as deprecated. To use data
                      from scan summary of numbers of hosts up, the user of the
                      lib will have to use NmapParser.parse() and the appropriate
                      accessors.
v0.4.1, 26/12/2013 -- Bug corrected: issue6
                      Infinite loop while launching several nmap scans in background

nmap_report.py
#!/usr/bin/env python 
from xml.dom import pulldom

class NmapReport:
    def __init__(self, nmap_report=None, type='XML'):
        self._nmap_xml_report = nmap_report
        self._nmap_dict_report = {}

    def parse_nmap_event(self, nmap_event):
        domevent = pulldom.parseString(nmap_event)
        for e, node in domevent:
            if event == pulldom.START_ELEMENT and node.tagName == 'p':
        print(node.toxml())
        doc.exandNode(node)
        # Following statement prints node with all its children '<p>Some text <div>and more</div></p>'
        print(node.toxml())




{u'nmaprun': {'#text': u'\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n',
              u'@args': u'nmap -sT -sV -oX reports-samples/scan_ss.xml -vv localhost scanme.nmap.org',
              u'@scanner': u'nmap',
              u'@start': u'1361385132',
              u'@startstr': u'Wed Feb 20 19:32:12 2013',
              u'@version': u'5.51',
              u'@xmloutputversion': u'1.03',
              u'debugging': {u'@level': u'0'},
              u'host': [{'#text': u'\n\n\n\n\n',
                         u'@endtime': u'1361385142',
                         u'@starttime': u'1361385133',
                         u'address': {u'@addr': u'127.0.0.1',
                                      u'@addrtype': u'ipv4'},
                         u'hostnames': {'#text': u'\n\n\n',
                                        u'hostname': [{u'@name': u'localhost',
                                                       u'@type': u'user'},
                                                      {u'@name': u'localhost',
                                                       u'@type': u'PTR'}]},
                         u'ports': {'#text': u'\n\n\n\n\n\n',
                                    u'extraports': {'#text': u'\n\n',
                                                    u'@count': u'995',
                                                    u'@state': u'closed',
                                                    u'extrareasons': {u'@count': u'995',
                                                                      u'@reason': u'conn-refused'}},
                                    u'port': [{u'@portid': u'22',
                                               u'@protocol': u'tcp',
                                               u'service': {u'@conf': u'10',
                                                            u'@extrainfo': u'protocol 2.0',
                                                            u'@method': u'probed',
                                                            u'@name': u'ssh',
                                                            u'@product': u'OpenSSH',
                                                            u'@version': u'5.3'},
                                               u'state': {u'@reason': u'syn-ack',
                                                          u'@reason_ttl': u'0',
                                                          u'@state': u'open'}},
                                              {u'@portid': u'25',
                                               u'@protocol': u'tcp',
                                               u'service': {u'@conf': u'10',
                                                            u'@hostname': u' bouteille.localdomain',
                                                            u'@method': u'probed',
                                                            u'@name': u'smtp',
                                                            u'@product': u'Postfix smtpd'},
                                               u'state': {u'@reason': u'syn-ack',
                                                          u'@reason_ttl': u'0',
                                                          u'@state': u'open'}},
                                              {u'@portid': u'111',
                                               u'@protocol': u'tcp',
                                               u'service': {u'@conf': u'10',
                                                            u'@extrainfo': u'rpc #100000',
                                                            u'@method': u'probed',
                                                            u'@name': u'rpcbind',
                                                            u'@version': u'2-4'},
                                               u'state': {u'@reason': u'syn-ack',
                                                          u'@reason_ttl': u'0',
                                                          u'@state': u'open'}},
                                              {u'@portid': u'631',
                                               u'@protocol': u'tcp',
                                               u'service': {u'@conf': u'10',
                                                            u'@method': u'probed',
                                                            u'@name': u'ipp',
                                                            u'@product': u'CUPS',
                                                            u'@version': u'1.4'},
                                               u'state': {u'@reason': u'syn-ack',
                                                          u'@reason_ttl': u'0',
                                                          u'@state': u'open'}},
                                              {u'@portid': u'3306',
                                               u'@protocol': u'tcp',
                                               u'service': {u'@conf': u'10',
                                                            u'@method': u'probed',
                                                            u'@name': u'mysql',
                                                            u'@product': u'MySQL',
                                                            u'@version': u'5.1.67'},
                                               u'state': {u'@reason': u'syn-ack',
                                                          u'@reason_ttl': u'0',
                                                          u'@state': u'open'}}]},
                         u'status': {u'@reason': u'localhost-response',
                                     u'@state': u'up'},
                         u'times': {u'@rttvar': u'1123',
                                    u'@srtt': u'2973',
                                    u'@to': u'100000'}},
                        {'#text': u'\n\n\n\n\n',
                         u'@endtime': u'1361385163',
                         u'@starttime': u'1361385133',
                         u'address': {u'@addr': u'74.207.244.221',
                                      u'@addrtype': u'ipv4'},
                         u'hostnames': {'#text': u'\n\n\n',
                                        u'hostname': [{u'@name': u'scanme.nmap.org',
                                                       u'@type': u'user'},
                                                      {u'@name': u'scanme.nmap.org',
                                                       u'@type': u'PTR'}]},
                         u'ports': {'#text': u'\n',
                                    u'extraports': {'#text': u'\n\n',
                                                    u'@count': u'1000',
                                                    u'@state': u'filtered',
                                                    u'extrareasons': {u'@count': u'1000',
                                                                      u'@reason': u'no-responses'}}},
                         u'status': {u'@reason': u'reset',
                                     u'@state': u'up'},
                         u'times': {u'@rttvar': u'5000',
                                    u'@srtt': u'367',
                                    u'@to': u'100000'}}],
              u'runstats': {'#text': u'\n',
                            u'finished': {u'@elapsed': u'30.65',
                                          u'@exit': u'success',
                                          u'@summary': u'Nmap done at Wed Feb 20 19:32:43 2013; 2 IP addresses (2 hosts up) scanned in 30.65 seconds',
                                          u'@time': u'1361385163',
                                          u'@timestr': u'Wed Feb 20 19:32:43 2013'},
                            u'hosts': {u'@down': u'0',
                                       u'@total': u'2',
                                       u'@up': u'2'}},
              u'scaninfo': {u'@numservices': u'1000',
                            u'@protocol': u'tcp',
                            u'@services': u'1,3-4,6-7,9,13,17,19-26,30,32-33,37,42-43,49,53,70,79-85,88-90,99-100,106,109-111,113,119,125,135,139,143-144,146,161,163,179,199,211-212,222,254-256,259,264,280,301,306,311,340,366,389,406-407,416-417,425,427,443-445,458,464-465,481,497,500,512-515,524,541,543-545,548,554-555,563,587,593,616-617,625,631,636,646,648,666-668,683,687,691,700,705,711,714,720,722,726,749,765,777,783,787,800-801,808,843,873,880,888,898,900-903,911-912,981,987,990,992-993,995,999-1002,1007,1009-1011,1021-1100,1102,1104-1108,1110-1114,1117,1119,1121-1124,1126,1130-1132,1137-1138,1141,1145,1147-1149,1151-1152,1154,1163-1166,1169,1174-1175,1183,1185-1187,1192,1198-1199,1201,1213,1216-1218,1233-1234,1236,1244,1247-1248,1259,1271-1272,1277,1287,1296,1300-1301,1309-1311,1322,1328,1334,1352,1417,1433-1434,1443,1455,1461,1494,1500-1501,1503,1521,1524,1533,1556,1580,1583,1594,1600,1641,1658,1666,1687-1688,1700,1717-1721,1723,1755,1761,1782-1783,1801,1805,1812,1839-1840,1862-1864,1875,1900,1914,1935,1947,1971-1972,1974,1984,1998-2010,2013,2020-2022,2030,2033-2035,2038,2040-2043,2045-2049,2065,2068,2099-2100,2103,2105-2107,2111,2119,2121,2126,2135,2144,2160-2161,2170,2179,2190-2191,2196,2200,2222,2251,2260,2288,2301,2323,2366,2381-2383,2393-2394,2399,2401,2492,2500,2522,2525,2557,2601-2602,2604-2605,2607-2608,2638,2701-2702,2710,2717-2718,2725,2800,2809,2811,2869,2875,2909-2910,2920,2967-2968,2998,3000-3001,3003,3005-3007,3011,3013,3017,3030-3031,3052,3071,3077,3128,3168,3211,3221,3260-3261,3268-3269,3283,3300-3301,3306,3322-3325,3333,3351,3367,3369-3372,3389-3390,3404,3476,3493,3517,3527,3546,3551,3580,3659,3689-3690,3703,3737,3766,3784,3800-3801,3809,3814,3826-3828,3851,3869,3871,3878,3880,3889,3905,3914,3918,3920,3945,3971,3986,3995,3998,4000-4006,4045,4111,4125-4126,4129,4224,4242,4279,4321,4343,4443-4446,4449,4550,4567,4662,4848,4899-4900,4998,5000-5004,5009,5030,5033,5050-5051,5054,5060-5061,5080,5087,5100-5102,5120,5190,5200,5214,5221-5222,5225-5226,5269,5280,5298,5357,5405,5414,5431-5432,5440,5500,5510,5544,5550,5555,5560,5566,5631,5633,5666,5678-5679,5718,5730,5800-5802,5810-5811,5815,5822,5825,5850,5859,5862,5877,5900-5904,5906-5907,5910-5911,5915,5922,5925,5950,5952,5959-5963,5987-5989,5998-6007,6009,6025,6059,6100-6101,6106,6112,6123,6129,6156,6346,6389,6502,6510,6543,6547,6565-6567,6580,6646,6666-6669,6689,6692,6699,6779,6788-6789,6792,6839,6881,6901,6969,7000-7002,7004,7007,7019,7025,7070,7100,7103,7106,7200-7201,7402,7435,7443,7496,7512,7625,7627,7676,7741,7777-7778,7800,7911,7920-7921,7937-7938,7999-8002,8007-8011,8021-8022,8031,8042,8045,8080-8090,8093,8099-8100,8180-8181,8192-8194,8200,8222,8254,8290-8292,8300,8333,8383,8400,8402,8443,8500,8600,8649,8651-8652,8654,8701,8800,8873,8888,8899,8994,9000-9003,9009-9011,9040,9050,9071,9080-9081,9090-9091,9099-9103,9110-9111,9200,9207,9220,9290,9415,9418,9485,9500,9502-9503,9535,9575,9593-9595,9618,9666,9876-9878,9898,9900,9917,9929,9943-9944,9968,9998-10004,10009-10010,10012,10024-10025,10082,10180,10215,10243,10566,10616-10617,10621,10626,10628-10629,10778,11110-11111,11967,12000,12174,12265,12345,13456,13722,13782-13783,14000,14238,14441-14442,15000,15002-15004,15660,15742,16000-16001,16012,16016,16018,16080,16113,16992-16993,17877,17988,18040,18101,18988,19101,19283,19315,19350,19780,19801,19842,20000,20005,20031,20221-20222,20828,21571,22939,23502,24444,24800,25734-25735,26214,27000,27352-27353,27355-27356,27715,28201,30000,30718,30951,31038,31337,32768-32785,33354,33899,34571-34573,35500,38292,40193,40911,41511,42510,44176,44442-44443,44501,45100,48080,49152-49161,49163,49165,49167,49175-49176,49400,49999-50003,50006,50300,50389,50500,50636,50800,51103,51493,52673,52822,52848,52869,54045,54328,55055-55056,55555,55600,56737-56738,57294,57797,58080,60020,60443,61532,61900,62078,63331,64623,64680,65000,65129,65389',
                            u'@type': u'connect'},
              u'taskbegin': [{u'@task': u'Ping Scan',
                              u'@time': u'1361385133'},
                             {u'@task': u'Parallel DNS resolution of 1 host.',
                              u'@time': u'1361385133'},
                             {u'@task': u'Connect Scan',
                              u'@time': u'1361385133'},
                             {u'@task': u'Service scan',
                              u'@time': u'1361385134'},
                             {u'@task': u'RPCGrind Scan against localhost (127.0.0.1)',
                              u'@time': u'1361385142'},
                             {u'@task': u'Connect Scan',
                              u'@time': u'1361385142'},
                             {u'@task': u'Service scan',
                              u'@time': u'1361385163'}],
              u'taskend': [{u'@extrainfo': u'1 total hosts',
                            u'@task': u'Ping Scan',
                            u'@time': u'1361385133'},
                           {u'@task': u'Parallel DNS resolution of 1 host.',
                            u'@time': u'1361385133'},
                           {u'@extrainfo': u'1000 total ports',
                            u'@task': u'Connect Scan',
                            u'@time': u'1361385133'},
                           {u'@extrainfo': u'5 services on 1 host',
                            u'@task': u'Service scan',
                            u'@time': u'1361385142'},
                           {u'@extrainfo': u'1 port',
                            u'@task': u'RPCGrind Scan against localhost (127.0.0.1)',
                            u'@time': u'1361385142'},
                           {u'@extrainfo': u'1000 total ports',
                            u'@task': u'Connect Scan',
                            u'@time': u'1361385163'}],
              u'verbose': {u'@level': u'2'}}}
